### PR TITLE
[persist] Consolidate opportunistically when fetching a part

### PR DESCRIPTION
**Pull request by bkirwi**
This requires splitting out a new `Cursor` abstraction, so that we can avoid holding on to a mutable pointer to `encoded_part` in `FetchedPart::next()`.

### Motivation

Did the work as part of #20708 prototyping, and it's a TODO in the code, so why not send the PR?

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/persist-client/src/fetch.rs [hunk 1]
```diff
@@ -493,14 +493,35 @@ where
...
+                }
+                if t != t_next {
+                    break;
+                }
+
+                // All equal... consolidate!
+                self.part_cursor.idx += 1;
+                d.plus_equals(&D::decode(d_next));
+            }
+
```
Comment (danhhz) [inline src/persist-client/src/fetch.rs:522]: can we do an `if d.is_zero() { continue; }` (and include coverage in the test)?

**Review by danhhz — APPROVED** [review]
cool!

File: src/persist-client/src/fetch.rs (hunk 1 above)
Comment (bkirwi) [reply src/persist-client/src/fetch.rs:522]: Yes!

**Comment by pH14** [comment]
Wow this is really cool

**Pull request merged**